packages/docprovider/src/notebookCellExecutor.ts

```diff
@@ -0,0 +1,168 @@
+    sessionDialogs,
+    translator
+  }: INotebookCellExecutor.IRunCellOptions): Promise<boolean> {
+    translator = translator ?? nullTranslator;
+    const trans = translator.load('jupyterlab');
+
+    switch (cell.model.type) {
+      case 'markdown':
+        (cell as MarkdownCell).rendered = true;
+        cell.inputHidden = false;
+        onCellExecuted({ cell, success: true });
+        break;
+      case 'code':
+        if (sessionContext) {
+          if (sessionContext.isTerminating) {
+            await showDialog({
+              title: trans.__('Kernel Terminating'),
+              body: trans.__(
+                'The kernel for %1 appears to be terminating. You can not run any cell for now.',
+                sessionContext.session?.path
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            break;
+          }
+          if (sessionContext.pendingInput) {
+            await showDialog({
+              title: trans.__('Cell not executed due to pending input'),
+              body: trans.__(
+                'The cell has not been executed to avoid kernel deadlock as there is another pending input! Submit your pending input and try again.'
+              ),
+              buttons: [Dialog.okButton()]
+            });
+            return false;
+          }
+          if (sessionContext.hasNoKernel) {
+            const shouldSelect = await sessionContext.startKernel();
+            if (shouldSelect && sessionDialogs) {
+              await sessionDialogs.selectKernel(sessionContext);
+            }
+          }
+
+          if (sessionContext.hasNoKernel) {
+            cell.model.sharedModel.transact(() => {
+              (cell.model as ICodeCellModel).clearExecution();
+            });
+            return true;
+          }
+
+          const kernelId = sessionContext?.session?.kernel?.id;
+          const apiURL = URLExt.join(
+            this._serverSettings.baseUrl,
+            `api/kernels/${kernelId}/execute`
+          );
+          const cellId = cell.model.sharedModel.getId();
+
+          // jupyverse case - it is undefined in jupyter-server
+          const fileId = notebook.sharedModel.getState('file_id') ?? '';
+          let documentId = `json:notebook:${fileId}`;
+          if (!fileId) {
+            if (
+              // FIXME sessionContext.path seems to be local - should we by-pass this test?
+              ['', this._drive.name].includes(
+                this._contents.driveName(
+                  sessionContext.session?.path ?? sessionContext.path
+                )
+              )
+            ) {
+              const localPath = this._contents.localPath(
+                sessionContext.session?.path ?? sessionContext.path
+              );
+              documentId =
+                this._drive.getRoomName({
+                  localPath: localPath,
+                  format: 'json',
+                  type: 'notebook'
+                }) ?? '';
+            }
```

Comment:
Thanks @krassowski for the review.

I thought about re-using the endpoint. But the issue is the overhead. It will be an overkill to call it at each execution request.

Alternatively, I could set up a cache based on a `WeakMap<NotebookModel, roomID>` and discard the entry if the execution is refused due to improper input parameters. What do you think